File: backend/app/core/upgrades/stat_modifiers/knockback_3.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class Knockback3(UpgradeImplementation):
    id = 'knockback_3'
    name = 'Knockback Boost'
    description = '+50% knockback'
    rarity = 'rare'
    type = 'stat_modifier'
    cost = 10

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 0.5
        appliers.modifiers.add_multiplier('attack', 'knockback', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/knockback_3.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const Knockback3Def: UpgradeDef = {
  id: "knockback_3",
  name: "Knockback Boost",
  description: "+50% knockback",
  rarity: RarityID.Rare,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 10,
  targetClass: UpgradeTargetID.Attack,
  fieldInTargetClass: UpgradeStatID.Knockback,
  value: 0.5,
  isMultiplier: true,
  stackable: true,
  maxStacks: 2,
}

export class Knockback3 extends Upgrade {}
